File: lesson_5/code/branching_process.m
% Parameters
numGenerations = 10;  % Number of generations to simulate
R0 = 1.5;             % Basic reproduction number
initialInfections = 1; % Number of initial infections

% Initialize arrays
numInfections = zeros(1, numGenerations);
numInfections(1) = initialInfections;

% Run the branching process
for gen = 2:numGenerations
    % Calculate the number of infections in the current generation
    currentInfections = 0;
    for i = 1:numInfections(gen-1)
        % Generate a Poisson-distributed number of new infections
        newInfections = poissrnd(R0);
        currentInfections = currentInfections + newInfections;
    end
    numInfections(gen) = currentInfections;
end

% Display the results
disp('Generation  Infections');
disp([(1:numGenerations)' numInfections']);

% Plot the results
figure;
bar(1:numGenerations, numInfections);
xlabel('Generation');
ylabel('Number of Infections');
title('Disease Spread using Branching Process');

File: lesson_5/code/branching_process.py
import matplotlib.pyplot as plt
import numpy as np


num_generations = 10
r0 = 1.5
initial_infections = 1

num_infections = np.zeros(num_generations, dtype=int)
num_infections[0] = initial_infections

for generation in range(1, num_generations):
    current_infections = 0
    for _ in range(num_infections[generation - 1]):
        current_infections += np.random.poisson(r0)
    num_infections[generation] = current_infections

print("Generation  Infections")
for generation, infections in enumerate(num_infections, start=1):
    print(generation, infections)

plt.bar(np.arange(1, num_generations + 1), num_infections)
plt.xlabel("Generation")
plt.ylabel("Number of Infections")
plt.title("Disease Spread using Branching Process")
plt.show()
